skyportal/facility_apis/sedm.py

```diff
@@ -0,0 +1,214 @@
+from . import FollowUpAPI
+from baselayer.app.env import load_env
+from datetime import datetime, timedelta
+import json
+import requests
+from requests_toolbelt.utils import dump
+
+env, cfg = load_env()
+
+
+def convert_request_to_sedm(request, method_value='new'):
+    photometry = sorted(request.obj.photometry, key=lambda p: p.mjd, reverse=True)
+    photometry_payload = {}
+
+    for p in photometry:
+        if (
+            p.filter.startswith('ztf')
+            and p.filter not in photometry_payload
+            and p.mag is not None
+        ):
+            # using filter[-1] as SEDM expects the bandpass name without "ZTF"
+            photometry_payload[p.filter[-1]] = {
+                'jd': p.mjd + 2_400_000.5,
+                'mag': p.mag,
+                'obsdate': p.iso.date().isoformat(),
+            }
+
+    rtype = request.payload['observation_type']
+    if rtype == 'IFU':
+        filters = ''
+        followup = 'IFU'
+    elif rtype == '3-shot (gri)':
+        filters = 'g,r,i'
+        followup = ''
+    elif rtype == '4-shot (ugri)':
+        filters = 'u,g,r,i'
+        followup = ''
+    elif rtype == '4-shot+IFU':
+        filters = 'u,g,r,i'
+        followup = 'IFU'
+    elif rtype == '3-shot+IFU':
+        filters = 'g,r,i'
+        followup = 'IFU'
+    elif rtype == "Mix 'n Match":
+        choices = request.payload['observation_choices']
+        hasspec = 'IFU' in choices
+        followup = 'IFU' if hasspec else ''
+        if hasspec:
+            choices.remove('IFU')
+        filters = ','.join(choices)
+    else:
+        raise ValueError('Cannot coerce payload into SEDM format.')
+
+    payload = {
+        'Filters': filters,
+        'Followup': followup,
+        'email': request.requester.contact_email,
+        'enddate': request.payload['end_date'],
+        'startdate': request.payload['start_date'],
+        'prior_photometry': photometry_payload,
+        'priority': request.payload['priority'],
+        'programname': request.allocation.group.name,
+        'requestid': request.id,
+        'sourceid': request.obj_id,
+        'sourcename': request.obj_id,
+        'status': method_value,
+        'username': request.requester.username,
+    }
+
+    return payload
+
+
+class SEDMAPI(FollowUpAPI):
+    """SkyPortal interface to the Spectral Energy Distribution machine (SEDM)."""
+
+    @staticmethod
+    def submit(request):
+        """Submit a follow-up request to SEDM."""
+        from ..models import DBSession, FollowupRequestHTTPRequest
+
+        payload = convert_request_to_sedm(request, method_value='new')
+        content = json.dumps(payload)
+        r = requests.post(
+            cfg['app.sedm_endpoint'], files={'jsonfile': ('jsonfile', content)},
+        )
+
```

Comment:
Should handle `requests.exceptions.ConnectionError`?

Reply:
This is handled in FollowupRequestHandler.